internal/querynode/segment.go

```diff
@@ -531,24 +540,40 @@ func (s *Segment) segmentDelete(offset int64, entityIDs *[]UniqueID, timestamps
 		           const long* primary_keys,
 		           const unsigned long* timestamps);
 	*/
+	if len(entityIDs) <= 0 {
+		return fmt.Errorf("empty pks to delete")
+	}
+
 	s.segPtrMu.RLock()
 	defer s.segPtrMu.RUnlock() // thread safe guaranteed by segCore, use RLock
 	if s.segmentPtr == nil {
 		return errors.New("null seg core pointer")
 	}
 
-	if len(*entityIDs) != len(*timestamps) {
+	if len(entityIDs) != len(timestamps) {
 		return errors.New("length of entityIDs not equal to length of timestamps")
 	}
 
 	var cOffset = C.int64_t(offset)
-	var cSize = C.int64_t(len(*entityIDs))
-	var cEntityIdsPtr = (*C.int64_t)(&(*entityIDs)[0])
-	var cTimestampsPtr = (*C.uint64_t)(&(*timestamps)[0])
-
-	status := C.Delete(s.segmentPtr, cOffset, cSize, cEntityIdsPtr, cTimestampsPtr)
-	if err := HandleCStatus(&status, "Delete failed"); err != nil {
-		return err
+	var cSize = C.int64_t(len(entityIDs))
+	var cTimestampsPtr = (*C.uint64_t)(&(timestamps)[0])
+
+	pkType := entityIDs[0].Type()
+	switch pkType {
+	case schemapb.DataType_Int64:
+		int64Pks := make([]int64, len(entityIDs))
+		for index, entity := range entityIDs {
+			int64Pks[index] = entity.(*int64PrimaryKey).Value
+		}
+		var cEntityIdsPtr = (*C.int64_t)(&int64Pks[0])
+		status := C.Delete(s.segmentPtr, cOffset, cSize, cEntityIdsPtr, cTimestampsPtr)
+		if err := HandleCStatus(&status, "Delete failed"); err != nil {
+			return err
+		}
+	case schemapb.DataType_VarChar:
+		//TODO::
```

Comment:
should we use the marshaled StringArray?

Reply:
agree